Allow external pytorch libraries to see kineto

astachowiczhabana opened the pull request:
Summary: External activity profilers needs access to kineto to register.
In scenario when profilers are implemented as a pytorch plugins it becomes problematic,
because kineto symbols are hidden.

Fixes #ISSUE_NUMBER


cc @malfet @seemethere

aaronenyeshi reviewed (approved):
LGTM

aaronenyeshi commented:
Out of curiosity, what external activity profilers are you using, and how are you using kineto?

astachowiczhabana commented:
Thank you @aaronenyeshi . This is part of Habana device enablement, new activities: https://github.com/pytorch/pytorch/pull/78645 https://github.com/pytorch/kineto/pull/611

astachowiczhabana commented:
CI errors should be fixed with https://github.com/pytorch/kineto/pull/612

astachowiczhabana committed:
Allow external pytorch libraries to see kineto

Summary: External activity profilers needs access to kineto to register.
In scenario when profilers are implemented as a pytorch plugins it becomes problematic,
because kineto symbols are hidden.

astachowiczhabana commented:
@aaronenyeshi Can we merge this pull request?

robieta commented:
Transitively exposing Kineto via PyTorch feels off to me, although @malfet would know better. It's also worth noting that Kineto isn't always available; we can't even rely on Kineto symbols in PyTorch core. (Although I'm working on adding limited visibility for things like enums.) We work around this by wrapping all the kineto calls we want to use and then handling the `#include <kineto.h>` and ugly `#ifdef`s in a cpp file. ([torch/csrc/profiler/kineto_shim.cpp](https://github.com/pytorch/pytorch/blob/master/torch/csrc/profiler/kineto_shim.cpp))

astachowiczhabana commented:
Hi @robieta and thank you for the review. I see your point. Actually I only need to access one kineto singleton from external plugin (LibkinetoApi& api() defined in third_party/kineto/libkineto/include/libkineto.h). This will allow to register hpu plugin as a custom profiling provider. If I understand you suggestion correctly I only need to define new symbol inside torch/csrc/profiler/kineto_shim.cpp  - something like get_libkineto_api(). And only that symbol will be visible and resolved by dynamic linker in runtime. Is this correct?

robieta commented:
@astachowiczhabana Exactly

astachowiczhabana committed:
Allow external pytorch libraries to see kineto

Summary: External activity profilers needs access to kineto to register.
In scenario when profilers are implemented as a pytorch plugins it becomes problematic,
because kineto symbols are hidden.

astachowiczhabana committed:
Merge branch 'master' into astachowicz_tb_hpu

astachowiczhabana commented:
@malfet @soulitzer can we merge this change?

robieta commented on torch/csrc/autograd/profiler_kineto.h:
```diff
@@ -382,6 +382,9 @@ TORCH_API void enableProfilerWithEventPostProcess(
 TORCH_API std::unique_ptr<ProfilerResult> disableProfiler();
 
+TORCH_API void registerProfilerFactory(
+    torch::profiler::impl::kineto::child_activity_profiler_factory_t factory);
```
`profiler_kineto.h` cannot directly include `kineto_shim.h` anymore. The reason is that we have started to make limited use of kineto headers in our internal implementation (just to get access to the enums) but that requires a strict separation of the implementation headers (like `kineto_shim.h`) and the API headers (like `profiler_kineto.h`). Otherwise kineto symbols leak out, things fail to link, and it's a mess. I apologize for the inconvenience; this changed since I first recommended this approach.

That said, I don't think that this is insurmountable. (The issue above was more about not being able to find function definitions.) If this is a one off we can probably just add some forward declares in this file and pass the factory through by pointer. On the other hand, if you're expecting to need more interaction with Kineto through pytorch we'll probably need to split `kineto_shim.h` in a "forward declares + pointers" header that is safe for the public API and a private only "actually include some kineto headers" header. WDYT?

astachowiczhabana replied:
Hi @robieta that's ok in this scenario, because only one symbol is required. Can you please take a look at the latest commit: only forward-declaring IActivityProfiler and creating separate factory function, no kineto_shim dependencies. Is this compatible with current code organization model?

robieta replied:
It's failing with:
```
error: could not convert 'factory' from 'torch::autograd::profiler::child_activity_profiler_factory_t {aka std::function<std::unique_ptr<libkineto::IActivityProfiler>()>}' to 'torch::profiler::impl::kineto::child_activity_profiler_factory_t {aka torch::profiler::impl::kineto::DummyChildActivityProfilerFactory}'
   torch::profiler::impl::kineto::registerProfilerFactory(factory);
```
Let me quickly see how painful the header split is. (I suspect not very.)

astachowiczhabana committed:
Merge branch 'pytorch:master' into astachowicz_tb_hpu

astachowiczhabana committed:
Allow external pytorch libraries to see kineto

Summary: External activity profilers needs access to kineto to register.
In scenario when profilers are implemented as a pytorch plugins it becomes problematic,
because kineto symbols are hidden.

astachowiczhabana committed:
Merge branch 'pytorch:master' into astachowicz_tb_hpu

robieta reviewed (approved):
LGTM. I'll take care of shepherding it through CI and landing. Thanks! (And looking forward to HPU profiling.)

astachowiczhabana committed:
Merge branch 'master' into astachowicz_tb_hpu

astachowiczhabana commented:
Hi @robieta. Any news in this topic? Can we merge this change?

kit1980 commented:
@robieta should we merge this?

robieta commented:
I never found the time to break up kineto_shim, unfortunately.

astachowiczhabana commented:
@kit1980  @robieta  Can we please merge this change?

robieta commented:
@pytorchbot rebase

robieta commented:
Looks like this is now too stale to auto rebase
